File: touch/src/main/java/de/nicidienase/chaosflix/touch/browse/ConferenceGroupFragment.java
package de.nicidienase.chaosflix.touch.browse;

import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import de.nicidienase.chaosflix.R;
import de.nicidienase.chaosflix.common.entities.recording.persistence.ConferenceGroup;
import de.nicidienase.chaosflix.touch.browse.adapters.ConferenceRecyclerViewAdapter;

public class ConferenceGroupFragment extends BrowseFragment {

	private static final String TAG = ConferenceGroupFragment.class.getSimpleName();

	private static final String ARG_COLUMN_COUNT = "column-count";
	private static final String ARG_GROUP = "group-name";
	private static final String LAYOUTMANAGER_STATE = "layoutmanager-state";
	private ConferencesTabBrowseFragment.OnConferenceListFragmentInteractionListener listener;

	private int columnCount = 1;
	private ConferenceGroup conferenceGroup;

	private ConferenceRecyclerViewAdapter conferencesAdapter;

	private RecyclerView.LayoutManager layoutManager;

	public ConferenceGroupFragment() {
	}

	public static ConferenceGroupFragment newInstance(ConferenceGroup group, int columnCount) {
		ConferenceGroupFragment fragment = new ConferenceGroupFragment();
		Bundle args = new Bundle();
		args.putInt(ARG_COLUMN_COUNT, columnCount);
		args.putParcelable(ARG_GROUP, group);
		fragment.setArguments(args);
		return fragment;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getArguments() != null) {
			columnCount = getArguments().getInt(ARG_COLUMN_COUNT);
			conferenceGroup = getArguments().getParcelable(ARG_GROUP);
		}

	}

	@Override
	public View onCreateView(LayoutInflater inflater, ViewGroup container,
							 Bundle savedInstanceState) {
		View view = inflater.inflate(R.layout.fragment_conferences_pager, container, false);

		if (view instanceof RecyclerView) {
			Context context = view.getContext();
			RecyclerView recyclerView = (RecyclerView) view;
			if (columnCount <= 1) {
				layoutManager = new LinearLayoutManager(context);
			} else {
				layoutManager = new GridLayoutManager(context, columnCount);
			}
			recyclerView.setLayoutManager(layoutManager);

			conferencesAdapter = new ConferenceRecyclerViewAdapter(listener);
			recyclerView.setAdapter(conferencesAdapter);
			getViewModel().getConferencesByGroup(conferenceGroup.getConferenceGroupId())
					.observe(this,conferenceList -> {
						conferencesAdapter.setItems(conferenceList);
						Parcelable layoutState = getArguments().getParcelable(LAYOUTMANAGER_STATE);
						if (layoutState != null) {
							layoutManager.onRestoreInstanceState(layoutState);
						}
					});
		}
		return view;
	}

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		if (context instanceof ConferencesTabBrowseFragment.OnConferenceListFragmentInteractionListener) {
			listener = (ConferencesTabBrowseFragment.OnConferenceListFragmentInteractionListener) context;
		} else {
			throw new RuntimeException(context.toString()
					+ " must implement OnListFragmentInteractionListener");
		}
	}

	@Override
	public void onSaveInstanceState(Bundle outState) {
		super.onSaveInstanceState(outState);
		outState.putParcelable(LAYOUTMANAGER_STATE, layoutManager.onSaveInstanceState());
	}

	@Override
	public void onPause() {
		super.onPause();
		getArguments().putParcelable(LAYOUTMANAGER_STATE, layoutManager.onSaveInstanceState());
	}

	@Override
	public void onDetach() {
		super.onDetach();
		listener = null;
	}
}

File: touch/src/main/java/de/nicidienase/chaosflix/touch/browse/ConferenceGroupFragment.kt
package de.nicidienase.chaosflix.touch.browse

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.Conference
import de.nicidienase.chaosflix.common.mediadata.entities.recording.persistence.ConferenceGroup
import de.nicidienase.chaosflix.common.viewmodel.BrowseViewModel
import de.nicidienase.chaosflix.common.viewmodel.ViewModelFactory
import de.nicidienase.chaosflix.touch.R
import de.nicidienase.chaosflix.touch.browse.adapters.ConferenceRecyclerViewAdapter
class ConferenceGroupFragment : Fragment() {

    private val viewModel: BrowseViewModel by viewModels { ViewModelFactory.getInstance(requireContext()) }

    private var layoutManager: RecyclerView.LayoutManager? = null

    private val columnCount: Int by lazy { resources.getInteger(R.integer.num_columns) }
    private val conferenceGroup: ConferenceGroup by lazy { arguments?.getParcelable<ConferenceGroup>(ARG_GROUP)!! }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_conferences_page, container, false)
        if (view is RecyclerView) {
            val context = view.getContext()
            layoutManager = if (columnCount <= 1) {
                LinearLayoutManager(context)
            } else {
                StaggeredGridLayoutManager(columnCount, StaggeredGridLayoutManager.VERTICAL)
            }
            view.layoutManager = layoutManager
            val conferencesAdapter = ConferenceRecyclerViewAdapter {
                findNavController().navigate(
                        ConferencesTabBrowseFragmentDirections.actionConferencesTabBrowseFragmentToEventsListFragment(conference = it)
                )
            }
            conferencesAdapter.setHasStableIds(true)
            view.adapter = conferencesAdapter
            viewModel.getConferencesByGroup(conferenceGroup.id).observe(viewLifecycleOwner, Observer<List<Conference>> { conferenceList: List<Conference>? ->
                if (conferenceList != null) {
                    conferencesAdapter.conferences = conferenceList
                    arguments?.getParcelable<Parcelable>(LAYOUTMANAGER_STATE)?.let {
                        layoutManager?.onRestoreInstanceState(it)
                    }
                }
            })
        }
        return view
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        layoutManager?.let {
            outState.putParcelable(LAYOUTMANAGER_STATE, it.onSaveInstanceState())
        }
    }

    override fun onPause() {
        super.onPause()
        layoutManager?.let {
            arguments?.putParcelable(LAYOUTMANAGER_STATE, it.onSaveInstanceState())
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        layoutManager = null
    }

    companion object {
        private val TAG = ConferenceGroupFragment::class.java.simpleName
        private const val ARG_COLUMN_COUNT = "column-count"
        private const val ARG_GROUP = "group-name"
        private const val LAYOUTMANAGER_STATE = "layoutmanager-state"
        @JvmStatic
        fun newInstance(group: ConferenceGroup?, columnCount: Int): ConferenceGroupFragment {
            val fragment = ConferenceGroupFragment()
            val args = Bundle()
            args.putInt(ARG_COLUMN_COUNT, columnCount)
            args.putParcelable(ARG_GROUP, group)
            fragment.arguments = args
            return fragment
        }
    }
}

